store/postgres/src/chain_store.rs

```diff
@@ -1813,6 +1841,120 @@ impl ChainStoreTrait for ChainStore {
     }
 }
 
+mod recent_blocks_cache {
+    use super::*;
+
+    type OffsetFromChainHead = BlockNumber;
+
+    struct Inner {
+        blocks: HashMap<OffsetFromChainHead, (BlockPtr, Option<json::Value>)>,
+        capacity: usize,
+    }
+
+    /// We cache the most recent blocks in memory to avoid overloading the
+    /// database with unnecessary queries close to the chain head. We invalidate
+    /// blocks:
+    /// - After 5 seconds.
+    /// - Whenever the chain head advances.
+    pub struct RecentBlocksCache {
+        // We protect everything with a global `RwLock` to avoid data races. Ugly...
+        inner: RwLock<Inner>,
+    }
+
+    impl RecentBlocksCache {
+        pub fn new(capacity: usize) -> Self {
+            RecentBlocksCache {
+                inner: RwLock::new(Inner {
+                    blocks: HashMap::new(),
+                    capacity,
+                }),
+            }
+        }
+
+        pub fn set_chain_head(&self, chain_head: BlockPtr) {
+            // We invalidate the entire cache when the chain head changes.
+            let mut inner = self.inner.write();
+            inner.blocks.clear();
+            inner.blocks.insert(0, (chain_head, None));
```

Comment:
This is overly aggressive, and will reduce the effectiveness of the cache; it should be enough to remove any entry that is not within `[chain_head - capacity + 1, chain_head]` which would help reduce the churn from this cache. In normal operation, we'd only need to remove one block from the cache when the chain head advances.

I also think it would be simpler if the cache didn't require a call to `set_chain_head` and instead infer that from `set_block`; that makes it less likely that future code changes forget to call this and make the cache ineffective.

Reply:
Oh .. I just realized that the hashmap is keyed off the distance from the chain head. If using a `HashMap`, why not just use the block number directly, even if that makes getting the current max block number a bit more expensive? If we want to go with `0 == chain_head`, then a `VecDeque` might be a better data structure.